deferred.ts

```diff
@@ -8,10 +8,10 @@ export type Deferred<T = any, R = Error> = {
 export type DeferredItemCreator<T> = () => Promise<T>;
 
 /** Create deferred promise that can be resolved and rejected by outside */
-export function defer<T>(): Deferred<T> {
+export function defer<T, R>(): Deferred<T> {
   let handled = false,
-    resolve,
-    reject;
+    resolve: (t?: T) => void | undefined,
+    reject: (r?: any) => void | undefined;
```

Comment:
`(r?: R)`?

Reply:
Promise rejections are a black hole in the typescript types ystem because of runtime errors.

with the proposed change: ```
error: Uncaught TypeError: error TS2322: Type '(r?: R | undefined) => void | undefined' is not assignable to type '(r?: Error | undefined) => void'.
  Types of parameters 'r' and 'r' are incompatible.
    Type 'Error | undefined' is not assignable to type 'R | undefined'.
      Type 'Error' is not assignable to type 'R'.
        'Error' is assignable to the constraint of type 'R', but 'R' could be instantiated with a different subtype of constraint '{}'.
```

Reply:
Okay, that's fine. Then I guess we'll need `// @ts-ignore` to suppress warning about `any`.